plugin-server/src/main/ingestion-queues/batch-processing/each-batch-buffer.ts

```diff
@@ -28,25 +28,29 @@ export async function eachBatchBuffer(
     }
     const batchStartTimer = new Date()
 
+    const messagesForLater: KafkaMessage[] = []
     let consumerSleepMs = 0
     for (const message of batch.messages) {
         // kafka timestamps are unix timestamps in string format
         const processAt = Number(message.timestamp) + queue.pluginsServer.BUFFER_CONVERSION_SECONDS * 1000
         const delayUntilTimeToProcess = processAt - Date.now()
 
-        if (delayUntilTimeToProcess < 0) {
+        if (delayUntilTimeToProcess <= 0) {
             await eachMessageBuffer(message, resolveOffset, queue)
         } else {
-            consumerSleepMs = Math.max(consumerSleepMs, delayUntilTimeToProcess)
+            if (consumerSleepMs < delayUntilTimeToProcess) {
+                consumerSleepMs = delayUntilTimeToProcess
+            }
+            messagesForLater.push(message)
         }
     }
-
-    // if consumerSleep > 0 it means we didn't process at least one message
-    if (consumerSleepMs > 0) {
+    if (messagesForLater.length > 0) {
         // pause the consumer for this partition until we can process all unprocessed messages from this batch
         await queue.bufferSleep(consumerSleepMs, batch.partition)
```

Comment:
How long could consumerSleepMs be? If it's long enough to trigger the consumer to be kicked out and rebalance on the kafka cluster, this could be a problem.

Another thing to think about: We currently consume some messages immediately, others only later. I think we shouldn't consume any messages in the batch until it's time to avoid double-processing.

Reply:
That was indeed possible. Added some regular heartbeat sending to avert his.

Reply:
Also made it so that we pause processing messages when the first message in need of sleep is encountered. This, along with addition of an extra `commitOffsetsIfNecessary()` should prevent double-processing.
Ultimately, I don't love any solution here, because Kafka is just not suited for processing events at a given time, but I think this should work.